### PR TITLE
Add earliest and latest compilers

### LINKED ISSUE #110: Add earliest/latest gcc and clang to Docker images
### Background

We currently run tests with whatever compiler is the default one on the platform, or sometimes we pick `gcc` or `clang` but not a specific version. The goal of this task is to set up an environment that has “earliest” and “latest” versions of the compilers. A complementary task arranges for us to run tests with those compilers.

We want “earliest” compilers to make sure that we aren't using features that didn't exist in those early compilers. Mostly, this is also enforced by compiling with options like `-pedantic` on a recent compiler, but we should also run at least one an actual job with an old compiler (also to check we aren't running into a compiler bug).

We want “latest” compilers mostly to check that the code compiles without warnings with recent compilers. Newer compilers tend to add more warnings.

### Goal of this task

* In the ubuntu-22.04 docker image, add [an apt package source from llvm.org](https://apt.llvm.org/). We'll only test with official versions, not with nightly builds of the compiler, so we only want the `llvm-toolchain-jammy-16` source and not the `llvm-toolchain-jammy` source. We'll update manually that to 17 when it's officially released and so on.
* In the ubuntu-22.04 docker image (the most recent platform we use), create symbolic links `/usr/local/bin/gcc-latest` and `/usr/local/bin/clang-latest` pointing to the most recent compiler versions available for that distribution. Edit the list of packages passed to `apt-get` to ensure that these versions get installed.. For gcc, just take `/usr/bin/gcc-12`. For clang, we'll use the packages from llvm.org — see previous bullet point — so currently it's `/usr/bin/clang-16`.
* In the ubuntu-16.04 docker image (the oldest platform we use), create symbolic links `/usr/local/bin/gcc-earliest` and `/usr/local/bin/clang-earliest`, pointing to the earliest version of the respective compilers that's available for that distribution: `/usr/bin/gcc-4.7` and `/usr/bin/clang-3.5`. Edit the list of packages passed to `apt-get` to ensure that these versions get installed.

### Validation

* Must run https://github.com/Mbed-TLS/mbedtls/issues/7891 correctly. (The new jobs don't need to pass, we just want them to run with the expected compiler versions, it's ok if the build still has errors.)
* Do it on both Jenkins instances since they each have their own cache of the Docker images. That's [internal CI](https://jenkins-mbedtls.oss.arm.com/) ([some internal documentation](https://confluence.arm.com/display/BSGSoftware/Continuous+Integration+%28CI%29+Systems)) and [OpenCI](https://ci.trustedfirmware.org/view/Mbed-TLS/).

### PR DESCRIPTION
**Pull request by gowthamsk-arm**
This commit updates the Ubuntu 16.04 docker file to include the earliest versions of GCC (4.7) and Clang (3.5) available in 16.04. For Ubuntu 22.04 the latest compiler versions of GCC (12) and Clang (16) are installed.

Also creates couple of symlinks of those compilers for simplified usage in the all.sh script.

`all.sh` will start using the new compiler versions in https://github.com/Mbed-TLS/mbedtls/pull/7968 and https://github.com/Mbed-TLS/mbedtls/pull/7969.

Resolves #110

### CODE REVIEW

**Comment by gilles-peskine-arm** [comment]
Test jobs on OpenCI:

* [development](https://mbedtls.trustedfirmware.org/job/mbedtls-release-ci-testing/2/)
* [mbedtls-2.28](https://mbedtls.trustedfirmware.org/job/mbedtls-release-ci-testing/3/)
* [use_earliest_latest_compilers](https://mbedtls.trustedfirmware.org/job/mbedtls-release-ci-testing/4/)
* [use_earliest_latest_compilers_2.28](https://mbedtls.trustedfirmware.org/job/mbedtls-release-ci-testing/5/)

File: resources/docker_files/ubuntu-16.04/Dockerfile [hunk 1]
```diff
@@ -112,7 +116,10 @@ RUN apt-get update -q && apt-get install -yq \
             libstdc++6:armhf \
             ;; \
     esac && \
-    rm -rf /var/lib/apt/lists/
+    rm -rf /var/lib/apt/lists/ && \
+    # create symbolic links for earliest gcc and clang versions
+    ln -s /usr/bin/gcc-4.7 /usr/bin/gcc-earliest && \
```
Comment (gilles-peskine-arm) [inline resources/docker_files/ubuntu-16.04/Dockerfile:121]: Please create the symlinks in `/usr/local/bin`. `/usr` is for what comes with the distribution. Also applies in the 22.04 dockerfile.

File: resources/docker_files/ubuntu-22.04/Dockerfile [hunk 2]
```diff
@@ -120,6 +120,22 @@ RUN apt-get update -q && apt-get install -yq \
     esac && \
     rm -rf /var/lib/apt/lists/
 
+RUN apt-get update -q && apt-get install -yq \
+    # provides some useful scripts for adding and removing repositories
+    software-properties-common && \
```
Comment (gilles-peskine-arm) [inline resources/docker_files/ubuntu-22.04/Dockerfile:125]: Please group all the installations from the standard repository in the same batch. It makes the file easier to read and the caching of intermediate steps more efficient.

File: resources/docker_files/ubuntu-22.04/Dockerfile [hunk 3]
```diff
@@ -120,6 +120,22 @@ RUN apt-get update -q && apt-get install -yq \
...
+    # add key for the apt package source
+    apt-key adv --keyserver hkp://keyserver.ubuntu.com:80 --recv 15CF4D18AF4F7421 && \
+    # add an apt package source from https://apt.llvm.org/ to install clang-16
+    add-apt-repository 'deb http://apt.llvm.org/jammy/ llvm-toolchain-jammy-16 main' -y && \
+    apt-get install -yq \
+    # to build Mbed TLS using latest clang version
+    clang-16 \
+    # to build Mbed TLS using latest gcc version
```
Comment (gilles-peskine-arm) [inline resources/docker_files/ubuntu-22.04/Dockerfile:133]: More precisely, latest gcc version available from Ubuntu.

File: resources/docker_files/ubuntu-22.04/Dockerfile [hunk 4]
```diff
@@ -120,6 +120,22 @@ RUN apt-get update -q && apt-get install -yq \
...
 
+RUN apt-get update -q && apt-get install -yq \
+    # provides some useful scripts for adding and removing repositories
+    software-properties-common && \
+    # add key for the apt package source
+    apt-key adv --keyserver hkp://keyserver.ubuntu.com:80 --recv 15CF4D18AF4F7421 && \
+    # add an apt package source from https://apt.llvm.org/ to install clang-16
+    add-apt-repository 'deb http://apt.llvm.org/jammy/ llvm-toolchain-jammy-16 main' -y && \
+    apt-get install -yq \
+    # to build Mbed TLS using latest clang version
```
Comment (gilles-peskine-arm) [inline resources/docker_files/ubuntu-22.04/Dockerfile:131]: More precisely, the latest clang version at the time of writing. (As opposed to: at the time the docker image is built, or at the time the tests run.)

**Review by gilles-peskine-arm — CHANGES_REQUESTED** [review]
A few minor change requests.

The test results look good, but we'll have to re-run at least some after updating the docker files.

**Comment by gowthamsk-arm** [comment]
Test jobs on OpenCI:

[development](https://mbedtls.trustedfirmware.org/job/mbedtls-release-ci-testing/20/) → PASS
[mbedtls-2.28](https://mbedtls.trustedfirmware.org/job/mbedtls-release-ci-testing/24/) → PASS
[use_earliest_latest_compilers](https://mbedtls.trustedfirmware.org/job/mbedtls-release-ci-testing/22/) → PASS
[use_earliest_latest_compilers_2.28](https://mbedtls.trustedfirmware.org/job/mbedtls-release-ci-testing/26/) → PASS

Test jobs on internal CI:

[use_earliest_latest_compilers](https://jenkins-mbedtls.oss.arm.com/job/mbedtls-release-ci-testing/519/) → PASS
[use_earliest_latest_compilers-2.28](https://jenkins-mbedtls.oss.arm.com/job/mbedtls-release-ci-testing/520/) → PASS

**Review by tgonzalezorlandoarm — APPROVED** [review]
LGTM

**Review by gilles-peskine-arm — CHANGES_REQUESTED** [review]
Just a few readability issues

**Comment by gilles-peskine-arm** [comment]
Note: in the future, please address review comments in new commits, rather than by pushing a new history, unless there's a good reason to overwrite the existing history.

File: resources/docker_files/ubuntu-22.04/Dockerfile [hunk 5]
```diff
@@ -100,6 +100,10 @@ RUN apt-get update -q && apt-get install -yq \
         zlib1g \
         # to build Mbed TLS with MBEDTLS_ZILIB_SUPPORT (removed in 3.0)
         zlib1g-dev \
+        # to build Mbed TLS using latest gcc version available from Ubuntu
+        gcc-12 \
```
[the remark this answers is not in the document]
Comment (gowthamsk-arm) [reply resources/docker_files/ubuntu-22.04/Dockerfile:104]: Will fix it.

File: resources/docker_files/ubuntu-22.04/Dockerfile [hunk 6]
```diff
@@ -118,7 +122,18 @@ RUN apt-get update -q && apt-get install -yq \
             libstdc++6:armhf \
             ;; \
     esac && \
-    rm -rf /var/lib/apt/lists/
+    rm -rf /var/lib/apt/lists/ 
+
+# For installing clang-16, we add the llvm package source and add corresponding key 
+RUN apt-key adv --keyserver hkp://keyserver.ubuntu.com:80 --recv 15CF4D18AF4F7421 \
+        # add an apt package source from https://apt.llvm.org/ 
+        &&  add-apt-repository 'deb http://apt.llvm.org/jammy/ llvm-toolchain-jammy-16 main' -y \
```
[the remark this answers is not in the document]
Comment (gowthamsk-arm) [reply resources/docker_files/ubuntu-22.04/Dockerfile:130]: Will change it.

**Review by gilles-peskine-arm — APPROVED** [review]
Looks good to me on code review. I'll check the CI logs when it's run on the latest release (no need to re-run all jobs, but at least build the image on both jenkins instances).

**Comment by gowthamsk-arm** [comment]
Thanks for the review :) 
I will trigger a job on both internal and external CIs.

**Comment by gowthamsk-arm** [comment]
Triggered. 
Internal CI run: https://jenkins-mbedtls.oss.arm.com/job/mbedtls-release-ci-testing/521/
External Ci run: https://mbedtls.trustedfirmware.org/job/mbedtls-release-ci-testing/27/

**Comment by gowthamsk-arm** [comment]
Both internal and external CI runs are green :)

**Pull request merged**